### ISSUE TITLE
[sklearn#34001] Make the _parameter_constraint attribute public and mandatory for every model

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#363]
**Mirrored from [scikit-learn#34001](https://github.com/scikit-learn/scikit-learn/issues/34001)**

Original title: Make the _parameter_constraint attribute public and mandatory for every model
Original labels: New Feature, Needs Decision, Developer API
Created: 2026-05-12T14:01:23Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the workflow you want to enable

I'm not sure if this has already been requested, but i haven't seen any discussion about this yet.

The `_parameter_constraint` attribute works for hyperparameter validation, but it can easily adapted to make hyperparameters available, and their ranges programatically accesible.

This would be great for any AutoML application that deals with hyperparameter optimization, since there would be an officially supported interface within which you can build the search space in which to optimize. To avoid making too many changes, we could expect this kind of workflows to filter parameters like `n_jobs`, `copy_X` or other internal parameters instead of making the distinction a the package level.  

### Describe your proposed solution

I see 2 possible ways to implement this:
1. Rename all `_parameter_constraint` attributes to `parameter_constraint`, making it clear that it is publicly accesible, and add references to it across the documentation pages.
2. Provide a new public method that returns the values in `_parameter_constaints`, adding a reference to it in the various documentation pages.

### Describe alternatives you've considered, if relevant

Users can already access `_parameter_constraints`. The only issue is that it's a private method and not well documented (which i understand, since now it's only meant for internal parameter validation).

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [48](https://github.com/godofecht/flow-scikit/issues/48) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [49](https://github.com/godofecht/flow-scikit/issues/49) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
